src/Controller/ControllerFactory.php

```diff
@@ -73,9 +96,23 @@ public function invoke($controller): ResponseInterface
         if ($result instanceof ResponseInterface) {
             return $result;
         }
-
         $action = $controller->getAction();
-        $args = array_values($controller->getRequest()->getParam('pass'));
+
+        $args = [];
+        $reflection = new ReflectionFunction($action);
+        $passed = array_values((array)$controller->getRequest()->getParam('pass'));
+        foreach ($reflection->getParameters() as $i => $parameter) {
+            if (isset($passed[$i])) {
+                $args[$i] = $passed[$i];
+                continue;
+            }
+            $class = $parameter->getClass();
+            $value = $parameter->getDefaultValue();
+            if ($class) {
+                $value = $this->container->get($class->getName());
```

Comment:
I think you should first check if the required name is available in container, using `has()`, before trying to `get()` it, else you will get an exception. (A test should be added for this case).

Reply:
Ok. I'll see what a more robust approach to handling parameters would be. I think you have a good point that injected services should be first in the parameter list as they are unlikely to be optional whilst passed routing parameters are more likely to be optional.